README.md
# Multiplayer-paint-tutorial

LAMP Stack Tutorial building a simple realtime multiplayer paint application.

The application is built following [TechLead tutorial](https://youtu.be/t1aXuJkmTg8).
Javascript, Firebase, PHP, Python, and Nginx are used for the tutorial.

The application is built only as a tutorial but not for other usages.

Thanks to TechLead for sharing an awesome tutorial.


CHANGELOG.MD
# Changelog
All notable changes to this project will be documented in this file.

## 2022-05-26
### Added
-  Allows editing of the existing drawing.


www/draw.php
<?php
/*
 * DEBUG 1 = show errors when they occur
 * DEBUG 0 = disable error reporting
 */
define('DEBUG', 1);
if (DEBUG) {
    ini_set('display_errors', '1');
    ini_set('display_startup_errors', '1');
    error_reporting(E_ALL);
}

$time = time();
$x = intval($_REQUEST['x']);
$y = intval($_REQUEST['y']);

// Save data
if (isset($_REQUEST['submit'])) {
        //print_r($_REQUEST);
        $data = $_POST['data'];
        $data = json_encode($data);

	$key = "$x,$y";
	//write data file outside of the document root
        $filename = "../tmp/$key";
	file_put_contents($filename, $data);
        $ret = null;
        $out = null;
        exec("python3 save.py 2>&1 '$x' '$y'", $out, $ret);
        if ($ret != 0) {
                print_r($out);
                die("Error saving($ret)");
	}

	// Redirect to homepage
        print "<script>window.location='index.php';</script>";

        return;
}

print <<<EOF
<!doctype html>
<html>
<body>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@simonwep/pickr/dist/themes/classic.min.css"/> <!-- 'classic' theme -->
<script src="draw.js?version=$time"></script>
<script src="https://cdn.jsdelivr.net/npm/@simonwep/pickr/dist/pickr.min.js"></script>

<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-app-compat.js"></script>
<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-firestore-compat.js"></script>
<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-auth-compat.js"></script>


<div id=picker></div>
<div>
<canvas id=mycanvas width=500 height=500 style="margin:8px;border:1px #000 solid"></canvas>
</div>

<!-- Save the x,y coordinates -->
<input type="submit" value = "Save" onclick="save($x, $y)">
<style>
@import url('https://fonts.cdnfonts.com/css/pixelsplitter');
html, body {
    height: 100%;
}

html {
    display: table;
    margin: auto;
    
}

body {
    display: table-cell;
    vertical-align: middle;
    //font-family: 'PixelSplitter', sans-serif;
}

input {
    font-family: 'PixelSplitter', sans-serif;
    padding: 5px;
}


</style>

</body>
</html>
EOF;
?>


www/index.js
$(document).ready(function () {
	let CANVAS = $("#mycanvas");
	let CTX = CANVAS.get(0).getContext("2d");
	let PIXELSIZE = 2;

	let REPEATSX = 20;
	let REPEATSY = 15;
	let DIMENSION = 25;

	let WIDTH = DIMENSION * REPEATSX * PIXELSIZE;
	let HEIGHT = DIMENSION * REPEATSY * PIXELSIZE;
	let SELECTEDBOX = null;

	CANVAS.attr('width', WIDTH);
	CANVAS.attr('height', HEIGHT);
        
	// Firebase credentials
	const firebaseConfig = {
		apiKey: "AIzaSyBJNeT_Z1GExU6LOMob2GfEKgQLeZIfZa4",
		authDomain: "bigcanvas-e50b1.firebaseapp.com",
 		projectId: "bigcanvas-e50b1",
 		storageBucket: "bigcanvas-e50b1.appspot.com",
 		messagingSenderId: "714000188961",
 		appId: "1:714000188961:web:79920d3d7c6483af644560"
 	};

 	// Initialize Firebase
 	const firebaseApp = firebase.initializeApp(firebaseConfig);
 	const db = firebaseApp.firestore();
 
 	db.collection('app').doc('grid').onSnapshot(function(doc) {
		// Read in the data
 		let data = doc.data();
 		for (let key in data) {
 			let coords = key.split(",");
 			let json = data[key];
 			let pixelData = JSON.parse(json);
 			for (let subkey in pixelData) {
 				let subcoord = subkey.split(",");
 				let color = pixelData[subkey];
 				fillPixel(coords, subcoord, color);
 			}
 		}
 	});

        function fillPixel(coords, subcoord, color) {
                let coordX = parseInt(coords[0]);
                let coordY = parseInt(coords[1]);
                let subcoordX = parseInt(subcoord[0]);
                let subcoordY = parseInt(subcoord[1]);

                CTX.fillStyle = color;
                let x = (coordX * DIMENSION + subcoordX) * PIXELSIZE;
                let y = (coordY * DIMENSION + subcoordY) * PIXELSIZE;
                CTX.fillRect(x, y, PIXELSIZE, PIXELSIZE);
        }
	
	CTX.strokeStyle = 'rgba(0,0,0,0.25';
	for (let i =0 ;i < DIMENSION * REPEATSX; ++i) {
		if (i % DIMENSION != 0) { continue; }
		x = i * PIXELSIZE;
		CTX.beginPath();
		CTX.moveTo(x, 0);
		CTX.lineTo(x, HEIGHT);
		CTX.stroke();

		y = i *PIXELSIZE;
		CTX.beginPath();
		CTX.moveTo(0, y);
		CTX.lineTo(WIDTH, y);
		CTX.stroke();
	}
	CANVAS.click(function(e) {
		selectBox(e);
	});

	// Track mousemove for selecting pixel
	CANVAS.mousemove(function(e){
		let pixel = [Math.floor(e.offsetX / (PIXELSIZE * DIMENSION)), Math.floor(e.offsetY / (PIXELSIZE * DIMENSION))];
		console.log(pixel);
		if(!SELECTEDBOX) {
			SELECTEDBOX = $("<div id=selectedBox></div");
			SELECTEDBOX.css({width: DIMENSION * PIXELSIZE -2, height: DIMENSION * PIXELSIZE -2})
			$("#mycanvasWrapper").prepend(SELECTEDBOX);
		}
		SELECTEDBOX.css({
			left: pixel[0] * PIXELSIZE * DIMENSION + 1,
			top: pixel[1] * PIXELSIZE * DIMENSION
		});
	});
        
	// Open drawing page of selected pixel
	let SELECTED = 0;
	function selectBox(e) {
		if (SELECTED) return;
		SELECTED = 1;

		let pixel = [Math.floor(e.offsetX / (PIXELSIZE * DIMENSION)), Math.floor(e.offsetY / (PIXELSIZE * DIMENSION))];
		window.location = "draw.php?x="+pixel[0]+"&y="+pixel[1];
	}

});


www/index.php
<?php
$time = time();

print <<<EOF
<!doctype html>
<html>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script src="index.js?version$time"></script>
<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-app-compat.js"></script>
<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-firestore-compat.js"></script>
<script src="https://www.gstatic.com/firebasejs/9.6.5/firebase-auth-compat.js"></script>

<style>
@import url('https://fonts.cdnfonts.com/css/pixelsplitter');
html, body {
    height: 100%;
}

html {
    display: table;
    margin: auto;
}

body {
    display: table-cell;
    vertical-align: middle;
    background-color: #f0f0f0;
    margin: 16px;
    font-family: 'PixelSplitter', sans-serif;
}
#mycanvas {
	border: 1px #000 solid;
	background-color: #fff;
	cursor:pointer;
	
}
#mycanvasWrapper {
	position:relative;
}
#selectedBox {
	border: 1px rgba(0, 50, 100, 0.5) solid;
	background-color: rgba(0, 50, 100, 0.25);
	position:absolute;
	pointer-events: none;
}

h1 {
	text-align: center;
	font-size: 50px;
}

h2 {
	text-align: center;
}

@keyframes bgcolor {
    0% {
        background-color: #ffcce6
    }

    50% {
        background-color: #ecccff
    }


    100% {
        background-color: #ccecff
    }
}

body {
    -webkit-animation: bgcolor 20s infinite;
    animation: bgcolor 10s infinite;
    -webkit-animation-direction: alternate;
    animation-direction: alternate;
}

</style>
<body>
<h1>Big Canvas</h1>
<h2>Click on an empty cell to show your creativity!</h2>
<div id=mycanvasWrapper>
<canvas id=mycanvas></canvas>
</div>
</body>
</html>
EOF;
?>


www/draw.js
function locationCoords() {
    let x = null;
    let y = null;

    urlStr = window.location.href;
    let params = urlStr.split('?')[1];
    let queryString = new URLSearchParams(params);

    for (let pair of queryString.entries()) {
        let key = pair[0];
        let val = pair[1];
        if (key === 'x')
            x = val;
        if (key === 'y')
            y = val;
    }

    return [x, y];
}

$(document).ready(function () {
	let CANVAS = $("#mycanvas");
	let CTX = CANVAS.get(0).getContext("2d");
	let DIMENSION = 25;
	let WIDTH = CANVAS.width();
	let HEIGHT = CANVAS.height();
	let PIXELSIZE = WIDTH / DIMENSION;
	let COLOR = '#000000';
	let ENABLED = true;
	let FILLED = {};
        
	// Canvas
	CTX.strokeStyle = 'rgba(0,0,0,0.1)';
	for (let i =0; i < DIMENSION; ++i) {
		x = Math.floor(i * WIDTH / DIMENSION);
		CTX.beginPath();
		CTX.moveTo(x, 0);
		CTX.lineTo(x, HEIGHT);
		CTX.stroke();

		y = Math.floor(i * HEIGHT / DIMENSION);
		CTX.beginPath();
		CTX.moveTo(0, y);
		CTX.lineTo(WIDTH, y);
		CTX.stroke();
	}
        
	// Fill pixel by the mousemove
        CANVAS.on('mousemove touchmove touchstart mousedown', mouseFill);
	function mouseFill(e) {
		if (!ENABLED) return;
		let offsetX = e.offsetX;
		let offsetY = e.offsetY;

	// Fill only when the mouse is pressed down	
		if (e.which != 1) return;
		pixel = [Math.floor(offsetX / PIXELSIZE), Math.floor(offsetY / PIXELSIZE)];
		fillPixel(pixel);
	}

	function fillPixel(pixel, color=COLOR){
		let key = pixel[0] + ',' + pixel[1];
                FILLED[key] = color;

		CTX.fillStyle = color;
		CTX.fillRect(pixel[0] * PIXELSIZE, pixel[1] * PIXELSIZE, PIXELSIZE - 1, PIXELSIZE - 1);
	}

// Color picker
const PICKR = Pickr.create({
    el: '#picker',
    theme: 'classic', // or 'monolith', or 'nano'
    comparison: false,
    components: {
	    opacity: false,
	    hue: true,
	    palette: true,
	    interaction: {
		    input: true,
	    }
    }
});
	PICKR.on('init', function () {
		PICKR.setColor(COLOR);
	});
	PICKR.on('show', function () {
		ENABLED = false;
	});
	PICKR.on('hide', function () {
		setTimeout(function () {
			ENABLED = true;
		}, 300);
	});
	PICKR.on('change', function () {
		COLOR = PICKR.getColor().toHEXA().toString();
	});
	
        // Save data
	window.save = function(x, y){
		var data = {};
		data['x'] = x;
		data['y'] = y;
		data['data'] = FILLED;

		//Send data to php script
		$.post('draw.php?submit=1', data, function(rsp) {
			$('body').append(rsp);
		});
	}
        //TODO modify code to populate draw canvas instead index canvas
        // Firebase credentials
	const firebaseConfig = {
		apiKey: "AIzaSyBJNeT_Z1GExU6LOMob2GfEKgQLeZIfZa4",
		authDomain: "bigcanvas-e50b1.firebaseapp.com",
 		projectId: "bigcanvas-e50b1",
 		storageBucket: "bigcanvas-e50b1.appspot.com",
 		messagingSenderId: "714000188961",
 		appId: "1:714000188961:web:79920d3d7c6483af644560"
 	};

 	// Initialize Firebase
 	const firebaseApp = firebase.initializeApp(firebaseConfig);
 	const db = firebaseApp.firestore();
 	let bigCoords = locationCoords();
 	db.collection('app').doc('grid').onSnapshot(function(doc) {
		// Read in the data
 		let data = doc.data();
			
		//Iterate over the big canvas coordinates
 		for (let key in data) {
 			let coords = key.split(",");
			if (bigCoords[0] != coords[0] || bigCoords[1] != coords[1]){
				continue; //next continue p
			}
			let json = data[key];
 			let pixelData = JSON.parse(json);

			//Iterate over each pixel color
 			for (let subkey in pixelData) {
 				let subcoord = subkey.split(",");
 				let color = pixelData[subkey];
 				fillPixel(subcoord, color);
 			}
 		}

 	});

});
